network/message/authorization.go

```diff
@@ -0,0 +1,364 @@
+package message
+
+import (
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/libp2p/message"
+	"github.com/onflow/flow-go/model/messages"
+	"github.com/onflow/flow-go/network/channels"
+)
+
+var authorizationConfigs map[string]MsgAuthConfig
+
+// MsgAuthConfig contains authorization information for a specific flow message. The authorization
+// is represented as a map from network channel -> list of all roles allowed to send the message on
+// the channel.
+type MsgAuthConfig struct {
+	// Name is the string representation of the message type.
+	Name string
+	// Type is a func that returns a new instance of message type.
+	Type func() interface{}
+	// Config is the mapping of network channel to list of authorized flow roles.
+	Config map[channels.Channel]flow.RoleList
+}
+
+// IsAuthorized checks if the specified role is authorized to send the message on the provided channel and
+// asserts that the message is authorized to be sent on the channel.
+// Expected error returns during normal operations:
+//  * ErrUnauthorizedMessageOnChannel: the channel is not included in the message's list of authorized channels
+//  * ErrUnauthorizedRole: the role is not included in the message's list of authorized roles for the provided channel
+func (m MsgAuthConfig) IsAuthorized(role flow.Role, channel channels.Channel) error {
```

Comment:
For methods with names `IsXYZ`, I would intuitively expect a **boolean** return value. I like your implementation, because the for the `false` case (role not authorized), we want to provide information which of the two conditions was violated. Thanks for   using the pattern with sentinel errors. Very nice. 

For me, a slightly different naming would be more intuitive:
```suggestion
func (m MsgAuthConfig) EnsureAuthorized(role flow.Role, channel channels.Channel) error {
```
I feel this would be more reflective of the implementation, where passing the method without any error return ensures that the node is authorized.

Reply:
https://github.com/onflow/flow-go/pull/2648/commits/3bc654fb3e844361d7f965e90487ef951876bdcf